libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -701,30 +701,40 @@ proc maintainDirectPeers(g: GossipSub) {.async.} =
     for id, addrs in g.parameters.directPeers:
       await g.addDirectPeer(id, addrs)
 
-method start*(g: GossipSub) {.async.} =
+method start*(
+    g: GossipSub
+): Future[void] {.async: (raises: [CancelledError], raw: true).} =
```

Comment:
why the raw here? can't see how it matters.

Reply:
not familiar enough to know _if_ it matters, but it makes the flow more explicit. there is no async involved in this method, so if the transformation introduces more complex logic or increases compile time, rather opting out of it

Reply:
makes it hard to read though - complexity is negligible in simple cases like this, specially since it's not part of any hot path

Reply:
yeah, that's true, the non-raw version is a bit simpler to read, but is a bit less clear about what's going on / as in, there is no async involved at all in the implementation, it's  just a regular synchronous function.

side note: one of the other PRs actually was triggered by switching some of the redirectors to `raw`, it changed test timing just slightly to make it fail, revealing a small logic issue.

anyway, will keep it in mind going forward and transform to `raw` a bit less agressively. I think one could argue that anytime `newFuture` can be avoided, the readability benefit generally outweighs the optimization aspect; while for redirections / wrappers where `newFuture` is not necessary, the `raw` version looks similar enough to the transformed version that there is no downside to using it?

can do another cleanup pass for readability after the raises annotations are done, depending on the outcome of the discussion. personally I'm also fine with `newFuture`.

Reply:
the thing with newFuture is that it opens up two avenues for "classic" bugs: returning nil on some code path _and_ returning an accidentally unfinished future on some code path - it's risky shit, for a dumb start function. ie when reviewing, I had to spend time ascertaining this point (it would have been more clear btw had you put the newfuture call at the end of the function where it gets returned, instead of at the beginning because as is, there's too much unnecessary distance between construction and use).

The most simple things are the most pleasurable to debate ;)

Reply:
True, haven't been bitten by those personally, but can see them happening relatively easily. It's a good point, generally, especially as it seems that the async transformation is not too costly for purely synchronous functions.